### PR TITLE
Fix float bug in help suggestion for incorrect tuple indexing

### PR DESCRIPTION
**Pull request by alexispurslane**
as per #28243.

### CODE REVIEW

**Comment by christopherdumas** [comment]
@nikomatsakis By the way, some changes from an earlier pull request of mine found their way into this PR branch, so I had to change them back.

**Comment by eddyb** [comment]
You should squash those commits to simplify the history.

**Comment by christopherdumas** [comment]
Yes, defiantly!

**Comment by nikomatsakis** [comment]
Can you add a testcase? I'm always torn about tests that have specific error msgs, but it seems appropriate in this instance. If you could include a comment identifying the precise thing you are testing (the formatting of the float in the suggestion) that'd be great.

**Comment by christopherdumas** [comment]
Sure.

**Comment by christopherdumas** [comment]
How do the tests work?

**Comment by christopherdumas** [comment]
I'm changing the help message in this PR, so it seems like I need to test the help message, but I'm not sure how using the tests that I see already there. Heres my start on the test (in `src/test/parse-fail/`):

```rust
// Copyright 2012 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution and at
// http://rust-lang.org/COPYRIGHT.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

// compile-flags: -Z parse-only

fn main () {
    (1, (2, 3)).1.1;
    //~ERROR unexpected token: `1.1`
}
```

**Comment by nikomatsakis** [comment]
@christopherdumas 

> I'm changing the help message in this PR, so it seems like I need to test the help message, but I'm not sure how using the tests that I see already there. 

I think you can just add a line like `//~ HELP ...` -- but the parse-fail tests may work differently than compile-fail? I forget. I'll have to check.

**Comment by christopherdumas** [comment]
Do you have any idea why this test is failing?
```rust
// Copyright 2012 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution and at
// http://rust-lang.org/COPYRIGHT.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

// compile-flags: -Z parse-only

fn main () {
    (1, (2, 3)).1.1; //~HELP try parenthesizing the first index; e.g., `(foo.1).1`
}

```

**Comment by jonas-schievink** [comment]
"Unexpected compiler error or warning". Looks like you have to tell it about the actual error. Try this:

```
(1, (2, 3)).1.1;  //~ ERROR unexpected token
                  //~^ HELP try parenthesizing the first index
```

The `//~^` tells it to look for the help message in the line above

**Comment by christopherdumas** [comment]
@jonas-schievink Thanks. I think that should do the trick!

**Comment by christopherdumas** [comment]
@nikomatsakis Can you make bors retry?

**Comment by eddyb** [comment]
@christopherdumas I don't see any approvals on this PR. The @bors message above was just a friendly reminder that the PR needed rebasing.

**Comment by christopherdumas** [comment]
Why is this a merge conflict? I tried to rebase to rust-lang/rust master, and llvm has a merge conflict. I tried all kinds of things, but I'm stuck.

**Comment by eddyb** [comment]
@christopherdumas You already fixed it, but the comment doesn't just go away.
GitHub tells you that "This branch is up-to-date with the base branch (Merging can be performed automatically)".

**Comment by eddyb** [comment]
@christopherdumas As for LLVM, use some tool like `git gui`, or at least never use `git commit -a`, to avoid committing changes to submodules like `src/llvm`.
You get those in between updating a branch and running `make` (which updates submodules as well).
Also, you should squash to hide those extra back-and-forth changes.

**Comment by christopherdumas** [comment]
@nikomatsakis This is ready to be merged....

File: src/test/parse-fail/tuple-float-index.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// compile-flags: -Z parse-only
+
+fn main () {
+    (1, (2, 3)).1.1; //~ ERROR unexpected token
+                     //~^ HELP try parenthesizing the first index
```
Comment (arielb1) [inline src/test/parse-fail/tuple-float-index.rs]: the actual fix would help (e.g. ````try parenthesizing the first index; e.g., `(foo.1).1````)

Comment (christopherdumas) [inline src/test/parse-fail/tuple-float-index.rs]: Totally! Thanks.

**Comment by arielb1** [comment]
Could you squash all commits together? r+ modulo that

**Comment by christopherdumas** [comment]
Yes! thanks.

**Comment by christopherdumas** [comment]
I squashed the commits.

**Comment by eddyb** [comment]
@christopherdumas You might want to amend the commit to fix the message (I don't think the issue there is even related to this PR).

**Comment by christopherdumas** [comment]
Whoops. sorry i must have gotten confused.

**Comment by christopherdumas** [comment]
I amended the commit message.

**Comment by eddyb** [comment]
Thanks! @bors r+

**Pull request merged**